waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,483 @@
+              ;; TODO - Look at events for messages about liveness probe failures:
+              ;; /api/v1/namespaces/<ns>/events?fieldSelector=involvedObject.namespace=<ns>,involvedObject.name=<instance-id>,reason=Unhealthy
+              ;; (-> event :message (string/starts-with? "Liveness probe failed:")) #{:never-passed-health-checks}
+              ;; For now, we just assume any SIGKILL (137) with the default "Error" reason was a livenessProbe kill.
+              killed-by-k8s? (and (= 137 (:exitCode newest-failure))
+                                  (= "Error" (:reason newest-failure)))
+              newest-failure-instance (merge live-instance
+                                             ; to match the behavior of the marathon scheduler,
+                                             ; don't include the exit code in failed instances that were killed by k8s
+                                             (when-not killed-by-k8s?
+                                               {:exit-code (:exitCode newest-failure)})
+                                             {:flags failure-flags
+                                              :healthy? false
+                                              :id (pod->instance-id pod (count failures))
+                                              :started-at newest-failure-start-time})]
+          (doseq [[id store] [[service-id service-id->failed-instances-transient-store]
+                              [instance-x instance-x->failed-instances-transient-store]]]
+            (swap! store update-in [id] (comp vec conj) newest-failure-instance)))))))
+
+(defn- pod->ServiceInstance
+  [pod]
+  ;; FIXME - catch exceptions and return nil
+  (let [port0 (get-in pod [:spec :containers 0 :ports 0 :containerPort])]
+  (scheduler/make-ServiceInstance
+    {:app-name (get-in pod [:metadata :labels :app])
+     :extra-ports (->> (get-in pod [:metadata :annotations :waiter/port-count])
+                       Integer/parseInt range next (mapv #(+ port0 %)))
+     :healthy? (get-in pod [:status :containerStatuses 0 :ready])
+     :host (get-in pod [:status :podIP])
+     :id (pod->instance-id pod)
+     :log-directory (str "/home/" (get-in pod [:metadata :namespace]))
+     :namespace (get-in pod [:metadata :namespace])
+     :pod-name (get-in pod [:metadata :name])
+     :port port0
+     :protocol (get-in pod [:metadata :annotations :waiter/protocol])
+     :restart-count (get-in pod [:status :containerStatuses 0 :restartCount])
+     :service-id (get-in pod [:metadata :annotations :waiter/service-id])
+     :started-at (-> pod
+                     (get-in [:metadata :creationTimestamp])
+                     (timestamp-str->datetime))})))
+
+(defn- api-request
+  [client url & {:keys [body content-type] :as options}]
+  (ss/try+
+    (log/info "K8s API request:" url options)
+    (let [auth-str @k8s-api-auth-str
+          result (pc/mapply mesos-utils/http-request client url
```

Comment:
Maybe rename `mesos-utils`? Or extract `http-request` to another file?

Reply:
#307